onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+{
+  bool modified = false;
+  for (Node& node : graph.Nodes()) {
+    if (node.OpType() == "Cast") {
+      const NodeAttributes& attributes = node.GetAttributes();
+      ORT_ENFORCE(attributes.find("to") != attributes.end());
+      bool is_fp = attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT);
+      bool is_fp16 = attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT16);
+      if (node.MutableOutputDefs().size() == 1) {
+        NodeArg* cast_output = node.MutableOutputDefs()[0];
+        for (Node* child : graph.GetMutableConsumerNodes(cast_output->Name())) {
+          if (child->OpType() == "Cast") {
+            const NodeAttributes& child_attributes = child->GetAttributes();
+            ORT_ENFORCE(child_attributes.find("to") != child_attributes.end());
+            bool is_child_fp = child_attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT);
+            bool is_child_fp16 = child_attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT16);
+            if ((is_fp && is_child_fp16) || (is_fp16 && is_child_fp)) {
+              // The parent and child cancell out
+              VLOGS(logger, 1) << "RemoveBackToBackCasts: Removed Cast nodes  " << node.Name() << " and " << child->Name();
+              RemoveCastNodes(graph, {&node, child});
+              modified = true;
+            } else if ((is_fp16 && is_child_fp16) || (is_fp && is_child_fp)) {
+              // Child is a duplicate of parent
+              VLOGS(logger, 1) << "RemoveBackToBackCasts: Removed Cast node  " << child->Name();
+              RemoveCastNodes(graph, {child});
+              modified = true;
+            }
+          }
+        }
+      }
+    }
+  }
+  return modified;
+}
+
+// SearchUpstream:
+// Recursively DFS traverse bottom-up, the graph upstream collecting all the NodeArgs that require a cast
+// inorder to move an FP16 Cast operation up the graph.
+// Visited float NodeArgs are either in require_cast or require_type_change so that the same
+// nodearg is traversed more than once.
+static void SearchUpstream(Graph& graph, NodeArg* node_arg,
+                           std::unordered_set<NodeArg*>& require_cast,
+                           std::unordered_set<NodeArg*>& require_type_change)
+{
+  Node* node = graph.GetMutableProducerNode(node_arg->Name());
+  if (node == nullptr) {
+    // The graph inputs don't have the producer nodes
+    if (node_arg->TypeAsProto()->tensor_type().elem_type() == TensorProto_DataType_FLOAT) {
+      require_cast.insert(node_arg);
+    }
+  } else {
+    std::string op_type = node->OpType();
+    if (op_type == "Cast" && node_arg->TypeAsProto()->tensor_type().elem_type() == TensorProto_DataType_FLOAT) {
+      // This Cast node and the Cast node that will be created later will cancel out
+      require_cast.insert(node_arg);
+    } else if (std::find(fp16_allow.begin(), fp16_allow.end(), op_type) == fp16_allow.end() &&
```

Comment:
Should we use a std::unordered_set or std::unordered_map for faster lookup ?

Reply:
and we should a utility function 

bool CanPropagate(string optypte, int input_index, int output_index) 
and it should check both allow_list